File: z字形变换.py
#z字形变换
'''
规律很巧妙，这题只能多举几个例子，找规律。
难找的是中间的数字的规律。
'''
class Solution:
    def convert(self, s, numRows):
        if not s:
            return ''
        if numRows == 1:
            return s
        length = len(s)
        cycleLen = 2 * numRows - 2  # 一个循环的个数
        result = ''
        for i in range(numRows):# i代表行数
            j = 0  # j用于在不同周期内跳转
            # j都是在那几个值变化0,cycleLen,2cycleLen...
            # 可以理解为每次都是在一个周期的开头
            # 而用j+i来控制在每个周期的位置
            # cycleLen-i表示在一个周期内，中间的那几个数字。
            while i + j < length:
                result += s[j + i]
                if i != 0 and i != numRows - 1 and j + cycleLen - i < length:
                    result += s[j + cycleLen - i]
                j += cycleLen
        return result

a=Solution()
print(a.convert('ABCDEFGHIJKLMN',2))

File: Z字形变换.cpp
#include <iostream>
#include <string>
#include <map>
#include <vector>
#include <set>
#include <algorithm>
#include <queue>
#include <unordered_set>
#include <time.h>

using namespace std;

class Solution {
public:
	string convert(string s, int numRows) {
		string res;
		if (numRows == 1)
			return s;
		int cycle = 2 * numRows - 2;//循环周期
		for (int i = 0; i < numRows; ++i)//用于控制Z字形中每一列
		{
			for (int j = 0; j + i < s.size(); j += cycle)//这里的j表示每个循环的开始
			{
				res += s[j + i];
				if (i != 0 && i != numRows - 1 && j + cycle - i < s.size())
					res += s[j + cycle - i];
			}
		}
		return res;
	}
};


int main()
{
	Solution a;

	string s = "LEETCODEISHIRING";
	cout << a.convert(s, 2) << endl;

	system("pause");
}

